README.md
# Snake Game with AI

## Usage

```
./snakeai [-chiqmnprst] [<prefix>brain-<score>.json]
  -c    create empty brain
  -h    start in human mode
  -i int
        max number of instances in epoch (default 1000)
  -m int
        min score in epoch
  -n float
        interval of the mutation changes on the synapse weight (default 0.5)
  -p string
        prefix filename with brain
  -q    start in silent mode
  -r float
        mutation rate on the weights of synapses (default 0.1)
  -s int
        snake speed limit (default 100)
  -t int
        max snake stept without eat (default 200)
```

```
$ go build -o snakeai
$ ./snakeai -c
$ ./snakeai brain-0.json
```

Terminal-based Snake game

![scrrenshot](http://i.imgur.com/pHf4fjt.gif)

## Play

### Locally

```
$ go get github.com/DyegoCosta/snake-game
$ $GOPATH/bin/snake-game
```

### On Docker

```
$ docker run -ti dyego/snake-game
```

## Testing

```
$ cd $GOPATH/src/github.com/DyegoCosta/snake-game
$ make
```


CHANGELOG.md
## [0.0.1]

### Added

-   Add ai


snake/game_test.go
package snake

import (
	"testing"
	"time"
)

func TestDefaultGameScore(t *testing.T) {
	g := NewGame()

	if g.score != 0 {
		t.Fatalf("Initial Game Score expected to be 0 but it was %d", g.score)
	}
}

func TestGameMoveInterval(t *testing.T) {
	e := time.Duration(85) * time.Millisecond
	g := NewGame()
	g.score = 150

	if d := g.moveInterval(100); d != e {
		t.Fatalf("Expected move interval to be %d but got %d", e, d)
	}
}

func TestAddPoints(t *testing.T) {
	g := NewGame()
	s := g.score
	g.addPoints(10)

	if s != 0 || g.score != 10 {
		t.Fatal("Expected 10 points to have been added to Game Score")
	}
}

func TestRetryGoBackToGameInitialState(t *testing.T) {
	g := NewGame()
	initScore := g.score
	initSnake := g.arena.snake

	g.arena.snake.changeDirection(UP)
	g.arena.moveSnake()
	g.addPoints(10)
	g.end()

	g.retry()

	if g.score != initScore {
		t.Fatal("Expected Score to have been reset")
	}

	for i, c := range g.arena.snake.body {
		if initSnake.body[i].x == c.x && initSnake.body[i].y == c.y {
			t.Fatal("Expected Snake body to have been reset")
		}
	}

	if g.arena.snake.direction == initSnake.direction {
		t.Fatal("Expected Snake direction to have been reset")
	}
}


snake/game.go
package snake

import (
	"os"
	"time"

	"github.com/imega/snake-game/state"
	"github.com/nsf/termbox-go"
)

var (
	pointsChan         = make(chan int)
	KeyboardEventsChan = make(chan KeyboardEvent)
)

// Game type
type Game struct {
	arena  *arena
	score  int
	isOver bool
}

func initialSnake() *snake {
	return newSnake(RIGHT, []coord{
		{x: 1, y: 1},
		{x: 1, y: 2},
		{x: 1, y: 3},
		{x: 1, y: 4},
	})
}

func initialScore() int {
	return 0
}

func initialArena() *arena {
	return newArena(initialSnake(), pointsChan, 20, 50)
}

func (g *Game) end() {
	g.isOver = true
}

func (g *Game) moveInterval(speed int) time.Duration {
	ms := 1*speed - (g.score / 10)
	return time.Duration(ms) * time.Millisecond
}

func (g *Game) retry() {
	g.arena = initialArena()
	g.score = initialScore()
	g.isOver = false
}

func (g *Game) addPoints(p int) {
	g.score += p
}

// NewGame creates new Game object
func NewGame() *Game {
	return &Game{arena: initialArena(), score: initialScore()}
}

// Start starts the game
func (g *Game) Start(p state.Parameters, ch chan state.SnakeGame, statCh chan state.Stat) {
	speed := p.Speed
	if err := termbox.Init(); err != nil {
		panic(err)
	}
	defer termbox.Close()

	go listenToKeyboard(KeyboardEventsChan)

	stat := state.Stat{}

	if err := g.render(p, stat); err != nil {
		panic(err)
	}

	go func() {
		for s := range statCh {
			stat = s
		}
	}()

	go func() {
		for e := range KeyboardEventsChan {
			switch e.EventType {
			case MOVE:
				d := keyToDirection(e.Key)
				g.arena.snake.changeDirection(d)
			case RETRY:
				g.retry()
			case SPEED:
				if e.Key == termbox.KeySpace {
					speed += 10
				}
				if e.Key == termbox.KeyBackspace {
					speed -= 10
				}
			case END:
				termbox.Close()
				os.Exit(0)
			}
		}
	}()

	for {
		select {
		case p := <-pointsChan:
			g.addPoints(p)
		default:
			if !g.isOver {
				if err := g.arena.moveSnake(); err != nil {
					g.end()
				}
			}

			var body []state.Coord
			for _, v := range g.arena.snake.body {
				body = append(body, state.Coord{
					X: v.x,
					Y: v.y,
				})
			}

			if err := g.render(p, stat); err != nil {
				panic(err)
			}

			if !p.Human {
				ch <- state.SnakeGame{
					Score:  g.score,
					IsOver: g.isOver,
					Arena: state.Arena{
						Width:  g.arena.width,
						Height: g.arena.height,
					},
					Food: state.Coord{
						X: g.arena.food.x,
						Y: g.arena.food.y,
					},
					Snake: state.Snake{
						Head: state.Coord{
							X: g.arena.snake.head().x,
							Y: g.arena.snake.head().y,
						},
						Body:  body,
						Steps: g.arena.snake.steps,
					},
				}
			}

			if speed > 0 {
				time.Sleep(g.moveInterval(speed))
			}
		}
	}
}


ai/neuronet.go
package ai

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/imega/snake-game/snake"
	"github.com/imega/snake-game/state"
	"github.com/nsf/termbox-go"
)

type Result struct {
	Neuronet neuronet
	Score    int
}

func New(p state.Parameters, ch chan state.SnakeGame, pad chan snake.KeyboardEvent, statCh chan state.Stat) error {
	var (
		population    []Result
		instance      int
		epoch         int
		maxEpochScore int
		bestBrain     Result

		maxInstance = p.MaxInstance

		n = neuronet{}

		lastState state.SnakeGame
		lastKey   termbox.Key
	)

	brain, err := loadBrain(p)
	if err != nil {
		return fmt.Errorf("failed to load brain, %s", err)
	}
	bestBrain = brain

	for st := range ch {
		var key termbox.Key

		if st.Snake.Steps > p.MaxSnakeSteps {
			pad <- snake.KeyboardEvent{EventType: snake.RETRY}

			instance++

			continue
		}

		if st.IsOver {
			if st.Score > p.MinScoreEpoch {
				population = append(population, Result{
					Neuronet: n,
					Score:    st.Score,
				})
			}
			if instance < maxInstance {
				pad <- snake.KeyboardEvent{EventType: snake.RETRY}
				n = mutate(bestBrain.Neuronet, p.MutationRate, p.MutationRange)

				instance++
			}

			if instance >= maxInstance {
				epoch++

				var best Result
				for _, p := range population {
					if best.Score < p.Score {
						best = p
					}
				}

				population = nil
				instance = 0
				maxEpochScore = 0

				if best.Score > 0 {
					bestBrain.Neuronet = crossoverBrain(bestBrain.Neuronet, best.Neuronet)
				}

				if bestBrain.Score < best.Score {
					bestBrain.Score = best.Score
					if err := saveBrain(p, bestBrain); err != nil {
						return fmt.Errorf("failed to save brain, %w", err)
					}
				}
			}

			if !p.Silent {
				statCh <- state.Stat{
					Epoch:         epoch,
					Instance:      instance,
					BestScore:     bestBrain.Score,
					MaxEpochScore: maxEpochScore,
				}
			}

			continue
		}

		if maxEpochScore < st.Score {
			maxEpochScore = st.Score
			if p.Silent {
				fmt.Printf(
					"Snake Game MaxScore: %d, epoch: %d, epochMaxScore: %d, inst: %d\n",
					bestBrain.Score,
					epoch,
					maxEpochScore,
					instance,
				)
			}
		}

		if lastState.Snake.Head.X == st.Snake.Head.X && lastState.Snake.Head.Y == st.Snake.Head.Y {
			continue
		}

		in := createInput(st)
		out := n.predict(in)

		var direction int
		var m float64
		for i, v := range out {
			if i == 0 || v > m {
				direction = i
				m = v
			}
		}

		switch direction {
		case 0:
			key = termbox.KeyArrowRight
		case 1:
			key = termbox.KeyArrowLeft
		case 2:
			key = termbox.KeyArrowUp
		case 3:
			key = termbox.KeyArrowDown
		}

		if lastKey != key {
			pad <- snake.KeyboardEvent{EventType: snake.MOVE, Key: key}
		}

		lastKey = key
	}

	return nil
}

type neuronet struct {
	WeightHidden1 [24][18]float64
	BiasHidden1   [18]float64
	WeightHidden2 [18][18]float64
	BiasHidden2   [18]float64
	WeightOutput  [18][4]float64
	BiasOut       [4]float64
}

func (n *neuronet) randFill() {
	rand.Seed(time.Now().UnixNano())

	for i := range n.WeightHidden1 {
		for j := range n.WeightHidden1[i] {
			n.WeightHidden1[i][j] = randFloat(-1, 1)
		}
	}

	for i := range n.BiasHidden1 {
		n.BiasHidden1[i] = randFloat(-1, 1)
	}

	for i := range n.WeightHidden2 {
		for j := range n.WeightHidden2[i] {
			n.WeightHidden2[i][j] = randFloat(-1, 1)
		}
	}

	for i := range n.BiasHidden2 {
		n.BiasHidden2[i] = randFloat(-1, 1)
	}

	for i := range n.WeightOutput {
		for j := range n.WeightOutput[i] {
			n.WeightOutput[i][j] = randFloat(-1, 1)
		}
	}

	for i := range n.BiasOut {
		n.BiasOut[i] = randFloat(-1, 1)
	}
}

func (n *neuronet) predict(input [24]float64) []float64 {
	var t1 [18]float64
	for i := range input {
		for j := range n.WeightHidden1[i] {
			if i == 0 {
				t1[j] += n.BiasHidden1[j]
			}
			t1[j] += input[i] * n.WeightHidden1[i][j]
		}
	}

	var h1 [18]float64
	for i := range t1 {
		h1[i] = ReLU(t1[i])
	}

	// Layer 2
	var t2 [18]float64
	for i := range h1 {
		for j := range n.WeightHidden2[i] {
			if i == 0 {
				t2[j] += n.BiasHidden2[j]
			}
			t2[j] += h1[i] * n.WeightHidden2[i][j]
		}
	}

	var h2 [18]float64
	for i := range t2 {
		h2[i] = ReLU(t2[i])
	}

	// OUT
	var t0 [4]float64
	for i := range h2 {
		for j := range n.WeightOutput[i] {
			if i == 0 {
				t0[j] += n.BiasOut[j]
			}
			t0[j] += h1[i] * n.WeightOutput[i][j]
		}
	}

	return SoftMax(t0[:])
}

func randFloat(min, max float64) float64 {
	rand.Seed(time.Now().UnixNano())

	return min + rand.Float64()*(max-min)
}

func ReLU(x float64) float64 {
	const (
		Overflow  = 1.0239999999999999e+03
		Underflow = -1.0740e+03
		NearZero  = 1.0 / (1 << 28) // 2**-28
	)

	switch {
	case math.IsNaN(x) || math.IsInf(x, 1):
		return x
	case math.IsInf(x, -1):
		return 0
	case x > Overflow:
		return math.Inf(1)
	case x < Underflow:
		return 0
	case -NearZero < x && x < NearZero:
		return 1 + x
	}

	if x > 0 {
		return x
	} else {
		return 0
	}
}

func SoftMax(x []float64) []float64 {
	var max float64 = x[0]
	for _, n := range x {
		max = math.Max(max, n)
	}

	a := make([]float64, len(x))

	var sum float64 = 0
	for i, n := range x {
		a[i] -= math.Exp(n - max)
		sum += a[i]
	}

	for i, n := range a {
		a[i] = n / sum
	}
	return a
}

func createInput(st state.SnakeGame) [24]float64 {
	in := [24]float64{}

	src := distanceHead2Arena(st)
	copy(in[:], src[:])

	src = distanceHead2Food(st)
	copy(in[8:], src[:])

	src = distanceHead2Body(st)
	copy(in[16:], src[:])

	return in
}

func distanceHead2Arena(st state.SnakeGame) [8]float64 {
	n := st.Arena.Height - st.Snake.Head.Y
	e := st.Arena.Width - st.Snake.Head.X
	s := st.Snake.Head.Y
	w := st.Snake.Head.X

	return distanceCardinalDirection(n, e, s, w)
}

func distanceHead2Food(st state.SnakeGame) [8]float64 {
	var n, e, s, w int
	if st.Snake.Head.X > st.Food.X {
		w = st.Snake.Head.X - st.Food.X
	} else {
		e = st.Food.X - st.Snake.Head.X
	}

	if st.Snake.Head.Y > st.Food.Y {
		s = st.Snake.Head.Y - st.Food.Y
	} else {
		n = st.Food.Y - st.Snake.Head.Y
	}

	return distanceCardinalDirection(n, e, s, w)
}

func distanceHead2Body(st state.SnakeGame) [8]float64 {
	var res [8]float64
	body := st.Snake.Body
	head := st.Snake.Head
	for i := range body {
		if head.X == body[i].X && head.Y < body[i].Y { // N ↑
			res[0] = float64(1) / float64(head.Y-body[i].Y)
		}

		if head.Y == body[i].Y && head.X < body[i].X { // E →
			res[1] = float64(1) / float64(body[i].X-head.X)
		}

		if head.X == body[i].X && head.Y < body[i].Y { // S ↓
			res[2] = float64(1) / float64(head.Y-body[i].Y)
		}

		if head.Y == body[i].Y && head.X > body[i].X { // W ←
			res[3] = float64(1) / float64(head.X-body[i].X)
		}

		if head.Y-body[i].Y == head.X-body[i].X && head.Y-body[i].Y > 0 { // SW ↙︎
			res[4] = float64(1) / float64(head.Y-body[i].Y)
		}

		if head.Y-body[i].Y == body[i].X-head.X && head.Y-body[i].Y > 0 { // SE ↘︎
			res[5] = float64(1) / float64(head.Y-body[i].Y)
		}

		if body[i].Y-head.Y == head.X-body[i].X && body[i].Y-head.Y > 0 { // NW ↖︎
			res[6] = float64(1) / float64(body[i].Y-head.Y)
		}

		if body[i].Y-head.Y == body[i].X-head.X && body[i].Y-head.Y > 0 { // NE ↗︎
			res[7] = float64(1) / float64(body[i].Y-head.Y)
		}
	}

	return res
}

func distanceCardinalDirection(n, e, s, w int) [8]float64 {
	var res [8]float64

	if n > 0 { // N ↑
		res[0] = float64(1) / float64(n)
	}

	if e > 0 { // E →
		res[1] = float64(1) / float64(e)
	}

	if s > 0 { // S ↓
		res[2] = float64(1) / float64(s)
	}

	if w > 0 { // W ←
		res[3] = float64(1) / float64(w)
	}

	if s > 0 && w > 0 { // SW ↙︎
		res[4] = float64(1 / math.Sqrt(float64(s^2+w^2)))
	}

	if s > 0 && e > 0 { // SE ↘︎
		res[5] = float64(1 / math.Sqrt(float64(s^2+e^2)))
	}

	if n > 0 && w > 0 { // NW ↖︎
		res[6] = float64(1 / math.Sqrt(float64(n^2+w^2)))
	}

	if n > 0 && e > 0 { // NE ↗︎
		res[7] = float64(1 / math.Sqrt(float64(n^2+e^2)))
	}

	return res
}

func mutate(n neuronet, mutationRate, mutationRange float64) neuronet {
	for i := range n.WeightHidden1 {
		for j := range n.WeightHidden1[i] {
			n.WeightHidden1[i][j] = mutateNeurone(n.WeightHidden1[i][j], mutationRate, mutationRange)
		}
	}

	for i := range n.BiasHidden1 {
		n.BiasHidden1[i] = mutateNeurone(n.BiasHidden1[i], mutationRate, mutationRange)
	}

	for i := range n.WeightHidden2 {
		for j := range n.WeightHidden2[i] {
			n.WeightHidden2[i][j] = mutateNeurone(n.WeightHidden2[i][j], mutationRate, mutationRange)
		}
	}

	for i := range n.BiasHidden2 {
		n.BiasHidden2[i] = mutateNeurone(n.BiasHidden2[i], mutationRate, mutationRange)
	}

	for i := range n.WeightOutput {
		for j := range n.WeightOutput[i] {
			n.WeightOutput[i][j] = mutateNeurone(n.WeightOutput[i][j], mutationRate, mutationRange)
		}
	}

	for i := range n.BiasOut {
		n.BiasOut[i] = mutateNeurone(n.BiasOut[i], mutationRate, mutationRange)
	}

	return n
}

func mutateNeurone(weight, mutationRate, mutationRange float64) float64 {
	if randFloat(0, 1) <= mutationRate {
		weight += randFloat(mutationRange*-1, mutationRange)
	}

	return weight
}

func saveBrain(p state.Parameters, bestBrain Result) error {
	b, err := json.Marshal(bestBrain.Neuronet)
	if err != nil {
		return fmt.Errorf("failed to unmarshal, %s", err)
	}

	prefix := ""
	if p.PrefixFilename != "" {
		prefix = p.PrefixFilename + "-"
	}

	f, err := os.Create(prefix + "brain-" + strconv.Itoa(bestBrain.Score) + ".json")
	if err != nil {
		return fmt.Errorf("failed to save file, %s", err)
	}

	if _, err := f.Write(b); err != nil {
		return fmt.Errorf("failed to write to file, %s", err)
	}

	if err := f.Close(); err != nil {
		return fmt.Errorf("failed to close file, %s", err)
	}

	return nil
}

func loadBrain(p state.Parameters) (Result, error) {
	var n neuronet

	nameWithScore := fileNameWithoutExtension(p.BrainFilename)

	pos := strings.LastIndexByte(nameWithScore, '-')
	if pos == -1 {
		return Result{}, fmt.Errorf("failed to parse filename. Need format <prefix>brain-<score>.json")
	}

	score := nameWithScore[pos+1:]

	b, err := ioutil.ReadFile(p.BrainFilename)
	if err != nil {
		return Result{}, fmt.Errorf("failed to read from file, %s", err)
	}

	if err := json.Unmarshal(b, &n); err != nil {
		return Result{}, fmt.Errorf("failed to marshal, %s", err)
	}

	s, err := strconv.Atoi(score)
	if err != nil {
		return Result{}, fmt.Errorf("failed to convert int from string, %s", err)
	}

	return Result{
		Neuronet: n,
		Score:    s,
	}, nil
}

func fileNameWithoutExtension(fileName string) string {
	if pos := strings.LastIndexByte(fileName, '.'); pos != -1 {
		return fileName[:pos]
	}

	return fileName
}

func crossoverBrain(n1, n2 neuronet) neuronet {
	next := n1

	for i := range n2.WeightHidden1 {
		if randFloat(0, 1) <= 0.5 {
			next.WeightHidden1[i] = n2.WeightHidden1[i]
		}
	}

	for i := range n2.BiasHidden1 {
		if randFloat(0, 1) <= 0.5 {
			next.BiasHidden1[i] = n2.BiasHidden1[i]
		}
	}

	for i := range n2.WeightHidden2 {
		if randFloat(0, 1) <= 0.5 {
			next.WeightHidden2[i] = n2.WeightHidden2[i]
		}
	}

	for i := range n2.BiasHidden2 {
		if randFloat(0, 1) <= 0.5 {
			next.BiasHidden2[i] = n2.BiasHidden2[i]
		}
	}

	for i := range n2.WeightOutput {
		if randFloat(0, 1) <= 0.5 {
			next.WeightOutput[i] = n2.WeightOutput[i]
		}
	}

	for i := range n2.BiasOut {
		if randFloat(0, 1) <= 0.5 {
			next.BiasOut[i] = n2.BiasOut[i]
		}
	}

	return next
}

func CreateBrain(p state.Parameters) error {
	n := neuronet{}
	n.randFill()

	r := Result{Neuronet: n}

	if err := saveBrain(p, r); err != nil {
		return fmt.Errorf("failed to save brain, %w", err)
	}

	return nil
}


snake/arena.go
package snake

import (
	"math/rand"
	"time"
)

type arena struct {
	food       *food
	snake      *snake
	hasFood    func(*arena, coord) bool
	height     int
	width      int
	pointsChan chan (int)
}

func newArena(s *snake, p chan (int), h, w int) *arena {
	rand.Seed(time.Now().UnixNano())

	a := &arena{
		snake:      s,
		height:     h,
		width:      w,
		pointsChan: p,
		hasFood:    hasFood,
	}

	a.placeFood()

	return a
}

func (a *arena) moveSnake() error {
	if err := a.snake.move(); err != nil {
		return err
	}

	if a.snakeLeftArena() {
		return a.snake.die()
	}

	if a.hasFood(a, a.snake.head()) {
		go a.addPoints(a.food.points)
		a.snake.steps = 0
		a.snake.length++
		a.placeFood()
	}

	a.snake.steps++

	return nil
}

func (a *arena) snakeLeftArena() bool {
	h := a.snake.head()
	return h.x > a.width || h.y > a.height || h.x < 0 || h.y < 0
}

func (a *arena) addPoints(p int) {
	a.pointsChan <- p
}

func (a *arena) placeFood() {
	var x, y int

	for {
		x = rand.Intn(a.width)
		y = rand.Intn(a.height)

		if !a.isOccupied(coord{x: x, y: y}) {
			break
		}
	}

	a.food = newFood(x, y)
}

func hasFood(a *arena, c coord) bool {
	return c.x == a.food.x && c.y == a.food.y
}

func (a *arena) isOccupied(c coord) bool {
	return a.snake.isOnPosition(c)
}


snake/snake.go
package snake

import "errors"

// Allowed snake movement directions
const (
	RIGHT direction = 1 + iota
	LEFT
	UP
	DOWN
)

type direction int

type snake struct {
	body      []coord
	direction direction
	length    int
	steps     int
}

func newSnake(d direction, b []coord) *snake {
	return &snake{
		length:    len(b),
		body:      b,
		direction: d,
	}
}

func (s *snake) changeDirection(d direction) {
	opposites := map[direction]direction{
		RIGHT: LEFT,
		LEFT:  RIGHT,
		UP:    DOWN,
		DOWN:  UP,
	}

	if o := opposites[d]; o != 0 && o != s.direction {
		s.direction = d
	}
}

func (s *snake) head() coord {
	return s.body[len(s.body)-1]
}

func (s *snake) die() error {
	return errors.New("Died")
}

func (s *snake) move() error {
	h := s.head()
	c := coord{x: h.x, y: h.y}

	switch s.direction {
	case RIGHT:
		c.x++
	case LEFT:
		c.x--
	case UP:
		c.y++
	case DOWN:
		c.y--
	}

	if s.isOnPosition(c) {
		return s.die()
	}

	if s.length > len(s.body) {
		s.body = append(s.body, c)
	} else {
		s.body = append(s.body[1:], c)
	}

	return nil
}

func (s *snake) isOnPosition(c coord) bool {
	for _, b := range s.body {
		if b.x == c.x && b.y == c.y {
			return true
		}
	}

	return false
}


snake/keyboard.go
package snake

import "github.com/nsf/termbox-go"

type keyboardEventType int

// Keyboard events
const (
	MOVE keyboardEventType = 1 + iota
	RETRY
	END
	SPEED
)

type KeyboardEvent struct {
	EventType keyboardEventType
	Key       termbox.Key
}

func keyToDirection(k termbox.Key) direction {
	switch k {
	case termbox.KeyArrowLeft:
		return LEFT
	case termbox.KeyArrowDown:
		return DOWN
	case termbox.KeyArrowRight:
		return RIGHT
	case termbox.KeyArrowUp:
		return UP
	default:
		return 0
	}
}

func listenToKeyboard(evChan chan KeyboardEvent) {
	termbox.SetInputMode(termbox.InputEsc)

	for {
		switch ev := termbox.PollEvent(); ev.Type {
		case termbox.EventKey:
			switch ev.Key {
			case termbox.KeyArrowLeft:
				evChan <- KeyboardEvent{EventType: MOVE, Key: ev.Key}
			case termbox.KeyArrowDown:
				evChan <- KeyboardEvent{EventType: MOVE, Key: ev.Key}
			case termbox.KeyArrowRight:
				evChan <- KeyboardEvent{EventType: MOVE, Key: ev.Key}
			case termbox.KeyArrowUp:
				evChan <- KeyboardEvent{EventType: MOVE, Key: ev.Key}
			case termbox.KeyEsc:
				evChan <- KeyboardEvent{EventType: END, Key: ev.Key}
			default:
				if ev.Ch == 'r' {
					evChan <- KeyboardEvent{EventType: RETRY, Key: ev.Key}
				}
			}
		case termbox.EventError:
			panic(ev.Err)
		}
	}
}


snake/presenter.go
package snake

import (
	"fmt"

	"github.com/imega/snake-game/state"
	"github.com/mattn/go-runewidth"
	"github.com/nsf/termbox-go"
)

const (
	defaultColor = termbox.ColorDefault
	bgColor      = termbox.ColorDefault
	snakeColor   = termbox.ColorGreen
)

func (g *Game) render(p state.Parameters, stat state.Stat) error {
	if p.Silent {
		return nil
	}

	termbox.Clear(defaultColor, defaultColor)

	var (
		w, h   = termbox.Size()
		midY   = h / 2
		left   = (w - g.arena.width) / 2
		right  = (w + g.arena.width) / 2
		top    = midY - (g.arena.height / 2)
		bottom = midY + (g.arena.height / 2) + 1
	)

	renderTitle(p, left, top, g.arena, stat)
	renderArena(g.arena, top, bottom, left)
	renderSnake(left, bottom, g.arena.snake)
	renderFood(left, bottom, g.arena.food)
	renderScore(left, bottom, g.score)
	renderQuitMessage(right, bottom)

	return termbox.Flush()
}

func renderSnake(left, bottom int, s *snake) {
	for _, b := range s.body {
		termbox.SetCell(left+b.x, bottom-b.y, ' ', snakeColor, snakeColor)
	}
}

func renderFood(left, bottom int, f *food) {
	termbox.SetCell(left+f.x, bottom-f.y, f.emoji, defaultColor, bgColor)
}

func renderArena(a *arena, top, bottom, left int) {
	for i := top; i < bottom; i++ {
		termbox.SetCell(left-1, i, '│', defaultColor, bgColor)
		termbox.SetCell(left+a.width, i, '│', defaultColor, bgColor)
	}

	termbox.SetCell(left-1, top, '┌', defaultColor, bgColor)
	termbox.SetCell(left-1, bottom, '└', defaultColor, bgColor)
	termbox.SetCell(left+a.width, top, '┐', defaultColor, bgColor)
	termbox.SetCell(left+a.width, bottom, '┘', defaultColor, bgColor)

	fill(left, top, a.width, 1, termbox.Cell{Ch: '─'})
	fill(left, bottom, a.width, 1, termbox.Cell{Ch: '─'})
}

func renderScore(left, bottom, s int) {
	score := fmt.Sprintf("Score: %v", s)
	tbprint(left, bottom+1, defaultColor, defaultColor, score)
}

func renderQuitMessage(right, bottom int) {
	m := "Press ESC to quit"
	tbprint(right-17, bottom+1, defaultColor, defaultColor, m)
}

func renderTitle(p state.Parameters, left, top int, a *arena, s state.Stat) {
	msg := "Snake Game in human mode"
	if !p.Human {
		msg = fmt.Sprintf(
			"Snake Game MaxScore: %d, epoch: %d, epochMaxScore: %d, inst: %d",
			s.BestScore,
			s.Epoch,
			s.MaxEpochScore,
			s.Instance,
		)
	}
	tbprint(left, top-1, defaultColor, defaultColor, msg)
}

func fill(x, y, w, h int, cell termbox.Cell) {
	for ly := 0; ly < h; ly++ {
		for lx := 0; lx < w; lx++ {
			termbox.SetCell(x+lx, y+ly, cell.Ch, cell.Fg, cell.Bg)
		}
	}
}

func tbprint(x, y int, fg, bg termbox.Attribute, msg string) {
	for _, c := range msg {
		termbox.SetCell(x, y, c, fg, bg)
		x += runewidth.RuneWidth(c)
	}
}


main.go
package main

import (
	"flag"
	"fmt"
	"os"

	"github.com/imega/snake-game/ai"
	"github.com/imega/snake-game/snake"
	"github.com/imega/snake-game/state"
)

func main() {
	p := state.Parameters{}

	flag.IntVar(&p.Speed, "s", 100, "snake speed limit")
	flag.IntVar(&p.MaxInstance, "i", 1000, "max number of instances in epoch")
	flag.Float64Var(&p.MutationRate, "r", 0.1, "mutation rate on the weights of synapses")
	flag.Float64Var(&p.MutationRange, "n", 0.5, "interval of the mutation changes on the synapse weight")
	flag.IntVar(&p.MaxSnakeSteps, "t", 200, "max snake stept without eat")
	flag.IntVar(&p.MinScoreEpoch, "m", 0, "min score in epoch")
	flag.StringVar(&p.PrefixFilename, "p", "", "prefix filename with brain")
	flag.BoolVar(&p.Silent, "q", false, "start in silent mode")
	flag.BoolVar(&p.Human, "h", false, "start in human mode")
	flag.BoolVar(&p.CreateBrain, "c", false, "create empty brain")
	flag.Parse()

	if p.CreateBrain {
		if err := ai.CreateBrain(p); err != nil {
			fmt.Printf("failed to create brain, %s\n", err)
			usage()
			os.Exit(1)
		}

		prefix := ""
		if p.PrefixFilename != "" {
			prefix = p.PrefixFilename + "-"
		}

		fmt.Printf("brain created: %sbrain-0.json\n", prefix)
		os.Exit(0)
	}

	ch := make(chan state.SnakeGame)
	statCh := make(chan state.Stat)

	if !p.Human {
		args := flag.Args()
		if len(args) == 0 {
			fmt.Printf("empty args filename\n")
			usage()
			os.Exit(1)
		}

		p.BrainFilename = args[0]

		go func() {
			if err := ai.New(p, ch, snake.KeyboardEventsChan, statCh); err != nil {
				fmt.Printf("failed to start, %s\n", err)
				usage()
				os.Exit(1)
			}
		}()
	}

	snake.NewGame().Start(p, ch, statCh)
}

func usage() {
	fmt.Fprintf(
		flag.CommandLine.Output(),
		"\nUsage: %s [-chiqmnprst] [<prefix>brain-<score>.json]\n",
		os.Args[0],
	)
	flag.PrintDefaults()
}
